src/Base/Tools.cpp

```diff
@@ -33,130 +33,217 @@
+    // Name should not already be there
+    assert(!baseNameAndDigitCountEntry.Contains(digitsValue));
+    baseNameAndDigitCountEntry.Add(digitsValue);
+}
+std::string Base::UniqueNameManager::makeUniqueName(const std::string& modelName,
+                                                    int minDigits) const
+{
+    std::string namePrefix;
+    std::string nameSuffix;
+    decomposeName(modelName, namePrefix, nameSuffix);
+    std::string baseName = namePrefix + nameSuffix;
+    auto baseNameEntry = UniqueSeeds.find(baseName);
+    if (baseNameEntry == UniqueSeeds.end()) {
+        // First use of baseName, just return it with no unique digits
+        return baseName;
+    }
+    // We don't care about the digit count of the suggested name, we always use at least the most
+    // digits ever used before.
+    int digitCount = baseNameEntry->second.size() - 1;
+    uint digitsValue;
+    if (digitCount < minDigits) {
+        // Caller is asking for more digits than we have in any registered name.
+        // We start the longer digit string at 000...0001 even though we might have shorter strings
+        // with larger numeric values.
+        digitCount = minDigits;
+        digitsValue = 1;
+    }
+    else {
+        digitsValue = baseNameEntry->second[digitCount].Next();
+    }
+    std::string digits = std::to_string(digitsValue);
+    if (digitCount > digits.size()) {
+        namePrefix += std::string(digitCount - digits.size(), '0');
+    }
+    return namePrefix + digits + nameSuffix;
 }
 
-std::string Base::Tools::addNumber(const std::string& name, unsigned int num, int d)
+// Remove a rgistered name so it can be generated again.
+// Nothing happens if you try to remove a non-registered name.
+void Base::UniqueNameManager::removeExactName(const std::string& name)
 {
-    std::stringstream str;
-    str << name;
-    if (d > 0) {
-        str.fill('0');
-        str.width(d);
+    std::string baseName;
+    std::string nameSuffix;
+    uint digitCount;
+    uint digitsValue;
+    std::tie(digitCount, digitsValue) = decomposeName(name, baseName, nameSuffix);
+    baseName += nameSuffix;
+    auto baseNameEntry = UniqueSeeds.find(baseName);
+    if (baseNameEntry == UniqueSeeds.end()) {
+        // name must not be registered, so nothing to do.
+        return;
+    }
+    int maxDigitCount = baseNameEntry->second.size();
+    if (digitCount >= maxDigitCount) {
+        // First use of this digitCount, name must not be registered, so nothing to do.
+        return;
+    }
+    PiecewiseSparseIntegerSet& baseNameAndDigitCountEntry = baseNameEntry->second[digitCount];
+    baseNameAndDigitCountEntry.Remove(digitsValue);
+    // Prune empty vector entries
+    while (--maxDigitCount >= 0 && !baseNameEntry->second[maxDigitCount].Any())
+        ;
+    if (maxDigitCount < 0) {
+        UniqueSeeds.erase(baseName);
+    }
+    else {
+        baseNameEntry->second.resize(maxDigitCount + 1);
     }
```

Comment:
My bad, the resizing makes it difficult to use `remove_if`.
I don't really like the use of while here, maybe `find_if` could be used instead? Something along the lines of:
```c++
auto& digitCountVector = baseNameEntry->second;
auto it = std::find_if(digitCountVector.rbegin(), digitCountVector.rend(),
                       [](const auto& set) { return set.Any(); });
if (it == digitCountVector.rend()) {
    UniqueSeeds.erase(baseName);
} else {
    digitCountVector.resize(std::distance(it, digitCountVector.rend()));
}

Reply:
`baseNameEntry->second` can contain empty (`!x.Any()`) entries below the non-empty one found at `maxDigitCount`, and those can't be removed because the index of the one that was at `maxDigitCount` will change.

Reply:
The issue isn't really that stl algorithms aren't used, but the cleaver while loop without a body. Is it correct way of using while? Sure! But it takes time to deduce what it does.

> Debugging code is twice as hard as writing the code in the first place. Therefore, if you write code as cleverly as possible, you are, by definition not smart enough to debug it.

--- Brian W. Kernighan

Maybe just something like this would be an improvement:
```c++
auto& digitCountVector = baseNameEntry->second;
const int maxDigitCount = digitCountVector.size();
//...
int newSize = maxDigitCount;
for (int i = maxDigitCount - 1; i >= 0; --i) {
    if (digitCountVector[i].Any()) {
        break;
    }
    newSize = i;
}
```

Or if that's too verbose, then maybe something like this could work?
```c++
auto& digitCountVector = baseNameEntry->second;
const int maxDigitCount = digitCountVector.size();
//...
int newSize = maxDigitCount;
for (int i = maxDigitCount - 1; i >= 0 && !digitCountVector[i].Any(); --i) {
    newSize = i;
}
```
By dumbing it down it might be easier to follow the code.

I like the following quote as a reminder to whom I write the code for:
> Any fool can write code that a computer can understand. Good programmers write code that humans can understand.

-- Martin Fowler

Reply:
The last two are constantly assigning to newSize, perhaps
```
int newSize = 0;
for (int i = maxDigitCount - 1; i >= 0; --i) {
    if (digitCountVector[i].Any()) {
        newSize = i;
        break;
    }
}
```

I frankly find all the suggestions no better than the original code. Does it take time to deduce what it does? Of course it does, all code does. It is a matter of whether the reader is accustomed to the paradigm and so is a matter of personal experience. Now, I'm an old fart who is not accustomed to the std algorithm libraries, or to C++ iterators in general (though I have C# enumerables and enumerators, and the linq library, down pat), so I find looping code easier to comprehend.

I feel one measure of comprehensibility would be how much documentation one has to read to understand what the code is doing. To understand the loop code one only needs to know basic C++ syntax and semantics and be familiar with the empty-loop concept where the effect of the code is entirely in the side effects of the looping condition, but to understand the std::find_if code one has to understand lambdas and several parts of the C++ std containers, iterators, and algorithms libraries. The more pre-knowledge one presumes of the reader, the more comprehensibility becomes subjective.

How about:
```
    do {
        --maxDigitCount;
    } while (maxDigitCount >= 0 && !baseNameEntry->second[maxDigitCount].Any());
```
which emphasizes that the point of the code is to reduce the value of `maxDigitCount`?

Perhaps the problem is the naming of `maxDigitCount` which actually starts off as one more than the maximum digit count (taken from `baseNameEntry->second.size()`) and only after this loop does it become the actual maximum count of digits.

The following consistently gives `maxDigitCount` the former meaning:
```
    while (maxDigitCount > 0 && !baseNameEntry->second[maxDigitCount-1].Any()) {
        --maxDigitCount;
    };
    if (maxDigitCount == 0) {
        UniqueSeeds.erase(baseName);
    }
    else {
        baseNameEntry->second.resize(maxDigitCount);
    }
```

whereas the following consistently gives `maxDigitCount` the latter meaning, which is the literal interpretation of its name:
```
    int maxDigitCount = baseNameEntry->second.size() - 1;
...
    while (maxDigitCount >= 0 && !baseNameEntry->second[maxDigitCount].Any()) {
        --maxDigitCount;
    };
    if (maxDigitCount < 0) {
        UniqueSeeds.erase(baseName);
    }
    else {
        baseNameEntry->second.resize(maxDigitCount + 1);
    }
```

Reply:
There are many benefits of using stl algorithms instead of raw for loops:
* for more complex cases it adds tested algorithms for common cases
* it removes a lot of potential issues that the boilerplate code needs to handle
* avoids repetition of common algorithms
* for simple cases it still shows the intent of the construct
* can parallelize execution

It is a bit verbose but that is going to change with the addition of ranges in c++20. When we get that we don't need to explicitly use the start and end iterators. When we get that (in april), we can do a regex change which fixes it.

I do agree that new stuff takes time to learn, and I'm starting to feel like an old fart my self when it comes to c++ and its new features, but the language is evolving and maybe we need to adjust.

As mentioned in the c++ practices document I linked to earlier that I base this review on (which still is in review and feedback is appreciated):
> Raw loops are those starting with `for`, `while` etc. While there are many options on how to write a loop, readabillity and maintainabillity should be the priority.

And in the code review document in same PR:
> This document aims to provide set of good practices that should be helpful for both developers and code reviewers. They should be treated like food recipes - you can play with them, alter them - but every change should be thoughtful and intentional.

I'm not trying to shoehorn stl functions in here, though I'm suggesting to think about if it can benefit the readability (and stability/maintainability) of the code for when the next person tries to patch it.
The code commented on in this case was, in my opinion, a bit difficult to read as the intent was not explicitly stated and it took some time to understand. I think your last two suggestions are much better but I still think it should be possible to improve it, but maybe it's not worth it. Pick the version you think is best. One last potential way of implementing it:
```c++
int newSize;
for (newSize = maxDigitCount - 1; newSize >= 0; --newSize) {
    if (digitCountVector[i].Any()) {
        break;
    }
}

if (newSize < 0) {
    UniqueSeeds.erase(baseName);
} 
else {
    baseNameEntry->second.resize(newSize + 1);
}
```

but it is up to you what you want to go with.